OpenTrons/opentrons-api
# fix(discovery-client): allow discovery client to track network interfaces

## overview

Previously, #5933 added network interface checking to enable Discovery Client restarts when the computer's active network interface list changed. This was required because:

- The discovery client's mDNS browser binds to network interfaces when it's created, so advertisements coming in on new interfaces would never be caught
- Plugging an OT-2 into the app via "USB" is actually an entire USB-to-Ethernet network interfaces

The implementation relied on the `app-shell` module to check the system's network interfaces, and if they changed, restart the entire discovery client. ~Either due to limitations in this implementation, or due to changes to the discovery client since then, this functionality either isn't working as well as we thought it was or has potentially broken entirely, as confirmed by real world testing.~

This worked, but only because the Discovery Client was also periodically re-sending its discovery query on an interval, which would ensure that once the DC picked up a new interface, it would (eventually) issue a discovery query on that interface. This was not an intentional effect of the re-query functionality, so when the re-query was removed according to #5985 and  #6193 added fairly aggressive discovery cache cleanup, robot rediscovery over the U2E interface became a lot less reliable.

This PR fixes the "hotplug" functionality, but on purpose this time. Related to #6585, in that this problem appears to be a root cause of the app never seeing an updated robot.

## changelog

Rather than rely on its caller, this PR pulls the network interface check directly into the Discovery Client in order to:

- Compare the system's network interface list directly with the interfaces that the mDNS browser has bound to
- Restart the mDNS browser rather than the entire Discovery Client on changes

It also:

- Reintroduces periodic rediscovery queries, but this time on an exponentially backed-off interval, topping out at once every 2 minutes

This brings the following benefits:

- As a standalone application / library, the Discovery Client can support U2E robot hotplugs on its own
    - This has very positive implications for stuff like factory QC tools and support tools
- Instead of restarting on "any" change, we now restart if we know the browser is not bound to known system interfaces (or if it's bound to too many)
    - The goal here is to restart more reliably and have some retry built in if we fail to bind to an interface on the first go or something 

## Review requests

Basic test procedure with Discovery Client CLI:

1. Make sure your app and OT-2 can _only_ talk over U2E, and is not plugged into the PC
    - e.g. disconnect your OT-2 from Wi-Fi
2. Build and start the discovery client CLI
    ```shell
    make -C discovery-client && yarn discovery browse --logLevel debug
    ```
3. Once the DC is running, plug the OT-2 in via U2E
4. Robot should eventually show up

Basic test procedure with app:

1. Make sure your app and OT-2 can _only_ talk over U2E, and is not plugged into the PC
    - e.g. disconnect your OT-2 from Wi-Fi
2. Clear the app's discovery cache with "More > Network & System > Clear Discovered Robots List"
3. Close the app and re-open via the command line so you can see the logs 
    ```shell
    make -C app dev
    ```
4. Wait a minute or so, and make sure the discovery spinner **is not running**
5. Watch for a log entry that looks like:
    - `debug: Restarting mDNS due to network interface mismatch { extra: [], missing: [ ... ] }`
6. The robot should show up as connectable in the list **without restarting discovery** via the button

## Risk assessment

Medium. There's a chance here that this mDNS browser restart strategy is overly aggressive in a way that could result in either excessive network load, missing valid advertisements due to frequent restarts, or both.

I've added app logging to record when these restarts happen, so we should carefully study these logs while testing and, if possible, stress test this change somehow.

If we saw the log happening every 5 seconds (its poll interval) on a given computer, that would be a red flag and force us to reconsider this approach.
